enclave-runner/src/usercalls/mod.rs

```diff
@@ -76,148 +83,307 @@ impl<T> Write for ReadOnly<T> {
     }
 }
 
-impl<W: Write> Write for WriteOnly<W> {
-    forward!(fn write(&mut self, buf: &[u8]) -> IoResult<usize>);
-    forward!(fn flush(&mut self) -> IoResult<()>);
+impl<T> AsyncWrite for ReadOnly<T> {
+    fn shutdown(&mut self) -> Poll<(), std::io::Error> {
+        return Err(IoErrorKind::BrokenPipe.into());
+    }
 }
 
-trait SharedStream<'a> {
-    type Inner: Read + Write + 'a;
+impl<T> AsyncRead for ReadOnly<T> where T: std::io::Read {}
 
-    fn lock(&'a self) -> Self::Inner;
+impl<T> Read for WriteOnly<T> {
```

Comment:
Move this and `impl<W: Write> Write for WriteOnly<W>` so it doesn't show up as changes in the diff.